central/auth/service/service_impl.go

```diff
@@ -89,3 +120,107 @@ func authStatusForID(id authn.Identity) (*v1.AuthStatus, error) {
 	}
 	return result, nil
 }
+
+func (s *serviceImpl) ListAuthMachineToMachineConfigs(ctx context.Context, _ *v1.Empty) (*v1.ListAuthMachineToMachineConfigResponse, error) {
+	storageConfigs, err := s.authDataStore.ListAuthM2MConfigs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.ListAuthMachineToMachineConfigResponse{Configs: storagetov1.AuthM2MConfigs(storageConfigs)}, nil
+}
+
+func (s *serviceImpl) GetAuthMachineToMachineConfig(ctx context.Context, id *v1.ResourceByID) (*v1.GetAuthMachineToMachineConfigResponse, error) {
+	config, exists, err := s.authDataStore.GetAuthM2MConfig(ctx, id.GetId())
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errox.NotFound.Newf("auth machine to machine config with id %q", id.GetId())
+	}
+	return &v1.GetAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(config)}, nil
+}
+
+func (s *serviceImpl) AddAuthMachineToMachineConfig(ctx context.Context, request *v1.AddAuthMachineToMachineConfigRequest) (*v1.AddAuthMachineToMachineConfigResponse, error) {
+	config := request.GetConfig()
+	if err := s.validateAuthMachineToMachineConfig(ctx, config, true); err != nil {
+		return nil, err
+	}
+	config.Id = uuid.NewV4().String()
+	storageConfig, err := s.authDataStore.AddAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config))
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.AddAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(storageConfig)}, nil
+}
+
+func (s *serviceImpl) UpdateAuthMachineToMachineConfig(ctx context.Context, request *v1.UpdateAuthMachineToMachineConfigRequest) (*v1.UpdateAuthMachineToMachineConfigResponse, error) {
+	config := request.GetConfig()
+	if err := s.validateAuthMachineToMachineConfig(ctx, config, false); err != nil {
+		return nil, err
+	}
+
+	storageConfig, err := s.authDataStore.UpdateAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config))
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.UpdateAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(storageConfig)}, nil
```

Comment:
RFC you linked in the [related thread](https://github.com/stackrox/stackrox/pull/8145#discussion_r1363675148) tells

>    If the target resource does not have a current representation and the PUT successfully creates one, then the origin server MUST inform the user agent by sending a 201 (Created) response.  If the target resource does have a current representation and that representation is successfully modified in accordance with the state of the enclosed representation, then the origin server MUST send either a 200 (OK) or a 204 (No Content) response to indicate successful completion of the request.

As you see, it is rather prescriptive on HTTP status codes (word `MUST`). Is there any possibility to detect if there was insertion or update and a possibility to set http status code for the response?

Reply:
Not really no, the store doesn't propagate that information in any way currently.

Reply:
Sadly. Let's leave this open for now and see if any of us has extra thoughts before it's time to merge the PR.

Reply:
Feel free to resolve.